Solutions/2.2-TapTweak-Solutions.ipynb

```diff
@@ -0,0 +1,647 @@
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# A) Generate Key Pair.\n",
+    "privatekey = ECKey()\n",
+    "privatekey.generate()\n",
+    "publickey = privatekey.get_pubkey()\n",
+    "\n",
+    "# B) Tweak: t < SECP256K1 Order\n",
+    "tweak = hashlib.sha256(b'tweak').digest()\n",
+    "assert(int.from_bytes(tweak, 'big') < SECP256K1_ORDER)\n",
+    "\n",
+    "# C) Tweak point.\n",
+    "tweak_private = ECKey()\n",
+    "tweak_private.set(tweak, True)\n",
+    "tweak_point = tweak_private.get_pubkey()\n",
+    "\n",
+    "# D) Derive tweaked private key.\n",
+    "privatekey_tweaked = privatekey.add(tweak)\n",
+    "\n",
+    "# E) Derive public key point.\n",
+    "publickey_tweaked = publickey.add(tweak_point)\n",
+    "\n",
+    "# F) Sign message with tweaked key pair and verify signature.\n",
+    "msg = hashlib.sha256(b'msg').digest()\n",
+    "sig = privatekey_tweaked.sign_schnorr(msg)\n",
+    "print(publickey_tweaked.verify_schnorr(sig, msg))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Programming Exercise 1.2 -  Signing with a tweaked 2-of-2 Musig key pairs.\n",
+    "\n",
+    "* A. 2-of-2 Musig key generation.\n",
+    "    * Key pair generation for all participants: `xG = P`\n",
+    "    * Aggregate all public keys: `pk_musig`\n",
+    "    * Apply challenge factors to all private keys: `x' = c * x`\n",
+    "* B. Tweak the `pk_musig`.\n",
+    "* C. Signing\n",
+    "    * Generate indidividual nonces & aggregate: `R_agg`\n",
+    "    * Sign individually & aggregate: `sig_agg`\n",
+    "    * _Question: Which participant(s) need to tweak keys?_\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# A1) Generate Key Pairs.\n",
+    "sec0 = ECKey()\n",
+    "sec1 = ECKey()\n",
+    "sec0.generate()\n",
+    "sec1.generate()\n",
+    "pk0 = sec0.get_pubkey()\n",
+    "pk1 = sec1.get_pubkey()\n",
+    "\n",
+    "# A2) Aggregate all public keys: \n",
+    "c_map, pk_musig = generate_musig_key([pk0, pk1])\n",
+    "\n",
+    "# A3) Apply challenge factors to keys.\n",
+    "sec0_c = sec0.mul(c_map[pk0])\n",
+    "sec1_c = sec1.mul(c_map[pk1])\n",
+    "pk0_c = pk0.mul(c_map[pk0])\n",
+    "pk1_c = pk1.mul(c_map[pk1])\n",
+    "    \n",
+    "# B) Tweak musig public key.\n",
+    "# Given: Tweak.\n",
+    "tweak = hashlib.sha256(b'tweak').digest()\n",
+    "pk_musig_tweaked = pk_musig.tweak_add(tweak)\n",
+    "\n",
+    "# C1) Nonce generation & aggregation.\n",
+    "k0 = generate_schnorr_nonce()\n",
+    "k1 = generate_schnorr_nonce()\n",
+    "R0 = k0.get_pubkey()\n",
+    "R1 = k1.get_pubkey()\n",
+    "R_agg, negated = aggregate_schnorr_nonces([R0, R1])\n",
+    "\n",
+    "# C2) Signing and signature aggregation.\n",
+    "msg = hashlib.sha256(b'msg').digest()\n",
+    "\n",
+    "# One person must tweak keys.\n",
```

Comment:
Technically you could even tweak the signature.
but yeah, if one of the participants tweak their keys that also works.
(don't know how to libsecp musig implementation will handle this)

Reply:
Let's leave this as one person tweaking their key when signing